### PR TITLE
Add architecture design of PACE framework projects' dependencies

### PR DESCRIPTION
**Pull request by mducle**
Adds a design document detailing the dependencies between the different components / projects of PACE.

Also includes options for removing / amalgamating the small `euphonic_horace` repository which acts as an interface between Euphonic and Horace.

And alternative names for some sub-projects / repositories.

Addresses: https://github.com/pace-neutrons/Pace-Project-Plan/issues/169

### CODE REVIEW

File: documentation/add/01_pace_framework.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# PACE framework
+
```
Comment (cmarooney-stfc) [inline documentation/add/01_pace_framework.md:2]: Add heading "COntents" or "Sections" or something

File: documentation/add/01_pace_framework.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+[Naming](#naming)
+
+
+## Introduction
+
+PACE comprises several independent programs which may interact with each other.
+This document defines how each program is expected to call the others, and especially the interfacing layers between the programs.
+
+
+## Dependencies
```
Comment (cmarooney-stfc) [inline documentation/add/01_pace_framework.md]: It is obvious from the diagram that MATLAB is sequestered in the top right corner. Whikle the individual labels convey the Python contents, a border round the Python modules marked Python in the same way as Matlab does would improve immediate comprehension . The pace-python distribution box is marked (various compiled stuff) but another label saying what unites them woud also be good.

File: documentation/add/01_pace_framework.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+![PACE project dependencies](diagrams/pace_framework.svg)
+
+The above diagram shows the dependencies of the various PACE projects.
+Arrows point from the dependent package to its dependency (e.g. `brillem` is dependent on `brille`).
+Solid arrows indicate a required dependency, whilst dashed arrows are optional dependencies
+(e.g. `pace-python` is only dependent on `euphonic_horace` for phonon calculations using Euphonic).
+
+Only projects which are publicly facing (where users can download a distributed package/toolbox/module) are shown.
+The two interface packages `euphonic_matlab` and `brillem` both depend on a separate library `light_python_wrapper`
+which they import as a git submodule.
```
Comment (cmarooney-stfc) [inline documentation/add/01_pace_framework.md]: DO we say here what light_python_wrapper does? If it is python, maybe outside the red matlab line? if it is a wrapper, what accesses its interface?

File: documentation/add/01_pace_framework.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+`light_python_wrapper` might be made a public Matlab toolbox at a later date.
+
+* The "Horace distribution" packages the `horace` and `herbert` repositories for Matlab users.
+  It is a program to visualise and analyse single crystal inelastic neutron spectra from time-of-flight spectrometers.
```
Comment (cmarooney-stfc) [inline documentation/add/01_pace_framework.md]: This looks like a brief description of Horace itself. Is it really necessary this far down the documentation? If you want to keep it, I would rephrase as "for Matlab users of Horace, the program to ..."

File: documentation/add/01_pace_framework.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+* The `pace-python` distribution is the equivalent of the "Horace distribution" aimed at Python users.
```
Comment (cmarooney-stfc) [inline documentation/add/01_pace_framework.md]: Well, not completely as it also exposes SPinW; it may be worth adding "users; itadditionally provides access to SpinW."

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]
@mducle Hi Duc - I'm having great problems with this. I've started doing comments below, but I think something more wholesale in the way of re-editing is required. The bullets are a very miscellaneous collection. Some of them are definitions of the main PACE modules - Horace/Herbert, SpinW, Euphonic, Brille - and I think these should go in a separate section - IF you think that is needed. They may be sufficiently well known to not be needed here. Some are definitions of the interfaces - euphonic-matlab/horace, brillem - and I think these should have pride of place as the main topic of this document, again in their own section. The diagram is not as clear as the one you sketched in the meeting. The packaging has got in the way of the connections. The labelling "toolbox" is obscuring for me what is connected to what. You may want briefly to say why you are using the toolbox label. The brille dependency is very secondary to the main goal of this text and maybe you could leave it out of the main diagram and put it in as a separate one. It is not clear to me what the interface code is in pace-python and whether that includes euphonic-matlab and/or brille. I think this should be unpacked. Happy to discuss further, maybe interactively, once you have done a first rewrite with these comments. Thanks Chris

File: documentation/add/01_pace_framework.md (hunk 2 above)
Comment (mducle) [reply documentation/add/01_pace_framework.md]: I've changed the labels to just say `Matlab` or `Python`.

File: documentation/add/01_pace_framework.md (hunk 3 above)
Comment (mducle) [reply documentation/add/01_pace_framework.md]: `light_python_wrapper` is a Matlab library which allows Matlab to access Python classes as if they are Matlab classes transparently.

File: documentation/add/01_pace_framework.md (hunk 4 above)
Comment (mducle) [reply documentation/add/01_pace_framework.md]: I've moved it to the bottom for people who don't know what it is...

File: documentation/add/01_pace_framework.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+## Naming
+
+Some project names, particularly `pace-python`, `euphonic_horace` and `euphonic_matlab` are thought to be lacking.
+Some suggested alternatives are:
+
+* For `pace-python`: `snakeskin`, `pacy`, `pace-neutron`
+* For `euphonic_horace`: `euphonic_horace_interface_python`, `euphonic_sqw`, `euphonic_horace_driver`
+* For `euphonic_matlab`: `euphonic_horace_interface_matlab`, `euphonium`, `meuphonic`
+
```
Comment (cmarooney-stfc) [inline documentation/add/01_pace_framework.md:144]: The discussion part of this document is dangling a bit. Ideally we would recrod decisions on euphonic-horace and on naming. For the moment I think it is acceptable to decide to leave things as-is now and revisit later e.g. after PACE, but we should record that decision here.

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]
Hi Duc - one change requested which is to record that we're not going to do any of the euphonic-horace or naming changes proposed, at the moment.

File: documentation/add/01_pace_framework.md (hunk 6 above)
Comment (mducle) [reply documentation/add/01_pace_framework.md:144]: Yeah... it petered out a bit... I've added a comment on the naming below this section.
